### ISSUE TITLE
Implementing off-grid sensors

### ISSUE THREAD

**Issue by tomelse** [ucl-bug/jwave#170]
Hi! I thought it would be really nice to introduce off-grid sensors into j-Wave, as they have done in k-Wave (http://www.k-wave.org/forum/topic/alpha-version-of-kwavearray-off-grid-sources). 

I've implemented a basic version of this myself using the band-limited interpolant, which you can see in this gist: https://gist.github.com/tomelse/f9ba7508b75f44f34ebdbf25d5f5b0a3. 

I'm happy to put a pull request in myself, I just wanted to get some feedback on how best to integrate it into the j-Wave code. 

![bli_comparison](https://github.com/ucl-bug/jwave/assets/56722586/30a42d55-ad4d-45fd-b5d1-593ee5ada008)

The results look really nice, removing all the staircasing that you can see in the on-grid version, which you can see in an example simulation above (x axis = time, y axis= detector). It's a little bit slower than the original j-Wave implementation but not prohibitively so (maybe that's to be expected?).

**Comment by astanziola** [comment]
Hello Thomas. That looks great, thanks a lot! Please do feel absolutely free to open a PR, it would be amazing to have off-grid sensors. 

I think this can easily go into the `geometry.py` file almost as is, although there may be a good argument to start having two separate `sources` and `sensors` modules.

I was looking at the comment given by @jgroehl below your gist and I think I agree with him. I can imagine a situation when someone want so do something like

```python
def my_loss_func(sensors_positions: Tuple[np.typing.ArrayLike]:
    sensors = BLISensors(sensors_positions)
    ...
    
sensor_positions_grad = jax.grad(my_loss_func)
```

In that way the position of the sensors can also be tuned by gradient descent! In that case, the flattening and unflattening methods also need to be changed in the pytree definition.

It would also be good to add some testing for this, for example by reproducing the followings:
- https://github.com/ucl-bug/k-wave/blob/2de03f2393e8a525798e0089cd15c04b033749a6/k-Wave/testing/unit/offGridPoints_test_on_grid.m
- https://github.com/ucl-bug/k-wave/blob/2de03f2393e8a525798e0089cd15c04b033749a6/k-Wave/testing/unit/offGridPoints_test_truncation.m

But the way the testing is handled in this repo currently is a bit clunky (any suggestion is more than welcome!), so I'm happy to add them afterwards 

Thank you for taking time to do this and share it!h

**Comment by astanziola** [comment]
Hello @tomelse ! Just gauging if you think you would like to make a PR on this? Otherwise I'm happy to take this over (of course acknowledging your contribution!)

**Comment by tomelse** [comment]
I'll have a look next week, had such a hectic week!

**Comment by tomelse** [comment]
Hi, 
I've been working on the implementation of this. I noticed that I hadn't dealt with the case that the sensor was precisely on the grid. This is my solution:

```python
def _bli_function(x0, x, n, include_imag=False):
    # x0 = position of sensors along axis
    # x = grid
    # n = size of grid
    # include_imag = include the imaginary component?
    dx = jnp.where((x - x0[:, None]) % 1 == 0, 1e-4 + x - x0[:, None], x - x0[:, None])
    if n % 2 == 0:
        y = jnp.sin(jnp.pi * dx) / \
            jnp.tan(jnp.pi * dx / n) / n
        y -= jnp.sin(jnp.pi * x0[:, None]) * jnp.sin(jnp.pi * x) / n
        if include_imag:
            y += 1j * jnp.cos(jnp.pi * x0[:, None]) * jnp.sin(jnp.pi * x) / n
    else:
        y = jnp.sin(jnp.pi * dx) / \
            jnp.sin(jnp.pi * dx / n) / n
    return y
```

It's a bit hacky, but I wanted to make sure that it was still differentiable close to the grid points. The original solution above didn't allow for that. 

Do you have any suggestions on better ways to do this? I thought about writing a custom derivative function but it seemed too complicated. 

I'm in the process of writing some tests - will upload soon!

**Comment by tomelse** [comment]
@astanziola I've implemented this in a new pull request (#192). There's example code, test code and an updated implementation.
